### PR TITLE
Black and import sorting

### PR DESCRIPTION
**Pull request by illusional**
- Import sorting with ruff
- Add `end-of-file-fixer` and `trailing-whitespace` to pre-commits
- Bump versions of pre-commits.

Note the non-standard line length of 120 for black / ruff here as @jmarshall and I discussed. In my testing, this did cause some dramas with out of date version of black in the environment that VSCode was using. Updating + restarting allowed this to work, but just an FYI.

This should be a no-op change.

We should aim to merge this ASAP, and I'll follow-up soon with enabling ruff linting (I've ignored just about all errors in the repo).

FYI @MattWellie as a heavy prod-pipes user.

### CODE REVIEW

**Comment by MattWellie** [comment]
Awesome, probably overdue :D 

Some of the file-end fixes look odd - there's a scrapping of final empty lines, which I thought end-of-file-fixer would enforce instead of strip 🤔

**Comment by illusional** [comment]
@MattWellie, can you just comment on one of the files, I'm diff blind 😓

File: .github/workflows/docker.yaml [hunk 1]
```diff
@@ -67,4 +67,4 @@ jobs:
         if: ${{ github.event_name == 'workflow_dispatch' && github.ref_name != 'main' }}
         run: |
           docker tag $IMAGE_NAME:${{ github.sha }} $DOCKER_DEV/$IMAGE_NAME:${{github.event.inputs.tag}}
-          docker push $DOCKER_DEV/$IMAGE_NAME:${{github.event.inputs.tag}}
\ No newline at end of file
+          docker push $DOCKER_DEV/$IMAGE_NAME:${{github.event.inputs.tag}}
```
Comment (MattWellie) [inline .github/workflows/docker.yaml:70]: e.g. @illusional

Comment (illusional) [reply .github/workflows/docker.yaml:70]: This is adding the EOF line?

Comment (MattWellie) [reply .github/workflows/docker.yaml:70]: Not sure I understand - the [plugin](https://github.com/pre-commit/pre-commit-hooks?tab=readme-ov-file#end-of-file-fixer) is described as 'Makes sure files end in a newline and only a newline.', and there's no newline here. I'm sure I'm wrong here, but not sure why.

Comment (illusional) [reply .github/workflows/docker.yaml:70]: GitHub shows the inverse, it puts the red <anchor> where there isn't a newline (see the left diff) and hides the newline where there is (right): 
image

Comment (MattWellie) [reply .github/workflows/docker.yaml:70]: how frustrating :D

**Review by vivbak — APPROVED** [review]
LGTM! Thanks @illusional 🙏 🙌

**Comment by MattWellie** [comment]
Is the plan to YOLO-merge this in? I've got a flake8 linting issue on another branch, I'm wondering whether to fix it or just rebase on top of this 👀

File: cpg_workflows/filetypes.py [hunk 2]
```diff
@@ -41,22 +42,18 @@ def __init__(
...
+            self.full_index_suffix = str(self.index_path).replace(str(self.path.with_suffix('')), '')
         self.reference_assembly = None
         if reference_assembly:
             self.reference_assembly = to_path(reference_assembly)
 
     @property
     @abstractmethod
-    def ext(self) -> str:
-        ...
+    def ext(self) -> str: ...
```
Comment (jmarshall) [inline cpg_workflows/filetypes.py:52]: I kind of wonder what it thinks it's doing here…

Could we avoid this by adding a docstring as is done for AlignmentInput.exists() above? e.g. (and ditto for `index_ext`)

```suggestion
    def ext(self) -> str:
        """The canonical extension for the file type, without a '.' at the start."""
```

File: cpg_workflows/jobs/align.py [hunk 3]
```diff
@@ -64,7 +62,7 @@ def _get_cram_reference_from_version(cram_version) -> str:
     if cram_version in cram_version_map:
         return cram_version_map[cram_version]
     raise ValueError(
-        f'Unrecognised cram_version: "{cram_version}", expected one of: {", ".join(cram_version_map.keys())}'
+        f'Unrecognised cram_version: "{cram_version}", expected one of: {", ".join(cram_version_map.keys())}',
```
Comment (jmarshall) [inline cpg_workflows/jobs/align.py:65]: I guess this is emblematic of my concerns with automated formatting…

We (and everyone else) only ever give ValueError one argument, but the multi-line formatting the autoformatter forces us to use means that we are obliged to give it a misleading trailing comma.

I had to look it up to check that this didn't turn the arguments into a tuple undesirably. (Turns out a trailing comma in a Python function call is a special case and doesn't turn the argument into a tuple.)

\<le big sigh>

Comment (illusional) [reply cpg_workflows/jobs/align.py:65]: I think this is a reasonable tradeoff for reduced diffs later on, same for lists, dicts, sets, etc.

File: cpg_workflows/jobs/align.py [hunk 4]
```diff
@@ -340,9 +326,7 @@ def _align_one(
 
     job_attrs = (job_attrs or {}) | dict(label=job_name, tool=aligner.name)
     if shard_number is not None and number_of_shards_for_realignment is not None:
-        job_name = (
-            f'{job_name} ' f'{shard_number + 1}/{number_of_shards_for_realignment} '
-        )
+        job_name = f'{job_name} ' f'{shard_number + 1}/{number_of_shards_for_realignment} '
```
Comment (jmarshall) [inline cpg_workflows/jobs/align.py:329]: Can we de-weird this while we're here…

```suggestion
        job_name = f'{job_name} {shard_number + 1}/{number_of_shards_for_realignment}'
```

File: cpg_workflows/jobs/align.py (hunk 3 above)
Comment (jmarshall) [reply cpg_workflows/jobs/align.py:65]: It's a reasonable tradeoff for things that do actually take a variable number of items.

It's suboptimal for things that take exactly one argument. Same with `dedent()` further on.

File: cpg_workflows/jobs/count.py [hunk 5]
```diff
@@ -37,33 +38,38 @@ class FeatureCounts:
...
+            '-t',
+            feature_type,
+            '-g',
+            attribute,
+            '-s',
+            {'none': '0', 'forward': '1', 'reverse': '2'}[strandness],
+            '-a',
+            str(gtf_file),
+            '-T',
+            str(threads),
```
Comment (jmarshall) [inline cpg_workflows/jobs/count.py:70]: Gah! Another example of "this semantically-blind autoformatting is not making this easier to read"

File: cpg_workflows/jobs/joint_genotyping.py [hunk 6]
```diff
@@ -53,10 +54,7 @@ def make_joint_genotyping_jobs(
     Outputs a multi-sample VCF under `output_vcf_path`.
     """
     if len(gvcf_by_sgid) == 0:
-        raise ValueError(
-            'Provided samples collection for joint calling should contain '
-            'at least one active sample'
-        )
+        raise ValueError('Provided samples collection for joint calling should contain ' 'at least one active sample')
```
Comment (jmarshall) [inline cpg_workflows/jobs/joint_genotyping.py:57]:
```suggestion
        raise ValueError('Provided samples collection for joint calling should contain at least one active sample')
```

File: cpg_workflows/jobs/trim.py [hunk 7]
```diff
@@ -18,11 +18,13 @@
 
 class MissingFastqInputException(Exception):
     """Raise if alignment input is missing"""
+
```
Comment (jmarshall) [inline cpg_workflows/jobs/trim.py:21]: (Insert standard "is this helping" rant here)

File: cpg_workflows/jobs/trim.py [hunk 8]
```diff
@@ -85,14 +87,20 @@ def __init__(
             raise ValueError(f'Invalid adapter type: {adapter_type}')
         self.command = [
             'cutadapt',
-            '-o', str(output_fastq_pair.r1),
-            '-a', adapters.r1.sequence,
+            '-o',
+            str(output_fastq_pair.r1),
+            '-a',
+            adapters.r1.sequence,
```
Comment (jmarshall) [inline cpg_workflows/jobs/trim.py:93]: This may be annoying enough that we should consider an annotation to suppress this reformatting for this statement.

Comment (illusional) [reply cpg_workflows/jobs/trim.py:93]: It's maybe more awkward, but you can reformat this as:

```python
[
    *('-o', str(output_fastq_pair.r1)),
    *('-a', adapters.r1.sequence),
    ...
]

File: cpg_workflows/large_cohort/combiner.py [hunk 9]
```diff
@@ -95,24 +86,21 @@ def run(
         params.setdefault('use_genome_default_intervals', True)
     else:
         raise ValueError(
-            'Either combiner/intervals must be set, or workflow/sequencing_type '
-            'must be one of: "exome", "genome"'
+            'Either combiner/intervals must be set, or workflow/sequencing_type ' 'must be one of: "exome", "genome"',
```
Comment (jmarshall) [inline cpg_workflows/large_cohort/combiner.py:89]:
```suggestion
            'Either combiner/intervals must be set, or workflow/sequencing_type must be one of: "exome", "genome"',
```

File: cpg_workflows/metamist.py [hunk 10]
```diff
@@ -429,50 +424,44 @@ def process_existing_analysis(
...
                 f'Not found completed analysis {label} for '
-                f'{f"sequencing group {sequencing_group_ids}" if len(sequencing_group_ids) == 1 else f"{len(sequencing_group_ids)} sequencing_groups" }'
+                f'{f"sequencing group {sequencing_group_ids}" if len(sequencing_group_ids) == 1 else f"{len(sequencing_group_ids)} sequencing_groups" }',
             )
         elif not completed_analysis.output:
-            logging.error(
-                f'Found a completed analysis {label}, '
-                f'but the "output" field does not exist or empty'
-            )
+            logging.error(f'Found a completed analysis {label}, ' f'but the "output" field does not exist or empty')
```
Comment (jmarshall) [inline cpg_workflows/metamist.py:430]:
```suggestion
            logging.error(f'Found a completed analysis {label}, but the "output" field does not exist or empty')
```

File: cpg_workflows/metamist.py [hunk 11]
```diff
@@ -429,50 +424,44 @@ def process_existing_analysis(
...
                 )
                 found_output_fpath = None
 
         # completed and good exists, can reuse
         if found_output_fpath:
-            logging.info(
-                f'Completed analysis {label} exists, '
-                f'reusing the result {found_output_fpath}'
-            )
+            logging.info(f'Completed analysis {label} exists, ' f'reusing the result {found_output_fpath}')
```
Comment (jmarshall) [inline cpg_workflows/metamist.py:449]:
```suggestion
            logging.info(f'Completed analysis {label} exists, reusing the result {found_output_fpath}')
```

File: cpg_workflows/metamist.py [hunk 12]
```diff
@@ -486,8 +475,7 @@ def process_existing_analysis(
         # proceeding with the standard workflow (creating status=queued, submitting jobs)
         else:
             logging.info(
-                f'Expected output file {expected_output_fpath} does not exist, '
-                f'so queueing analysis {label}'
+                f'Expected output file {expected_output_fpath} does not exist, ' f'so queueing analysis {label}',
```
Comment (jmarshall) [inline cpg_workflows/metamist.py:478]:
```suggestion
                f'Expected output file {expected_output_fpath} does not exist, so queueing analysis {label}',
```

File: cpg_workflows/metamist.py [hunk 13]
```diff
@@ -574,34 +560,27 @@ def parse_reads(  # pylint: disable=too-many-return-statements
...
-        raise MetamistError(
-            f'{sequencing_group_id}: no "meta/reads_type" field in meta'
-        )
+        raise MetamistError(f'{sequencing_group_id}: no "meta/reads_type" field in meta')
     supported_types = ('fastq', 'bam', 'cram')
     if reads_type not in supported_types:
         raise MetamistError(
-            f'{sequencing_group_id}: ERROR: "reads_type" is expected to be one of '
-            f'{supported_types}'
+            f'{sequencing_group_id}: ERROR: "reads_type" is expected to be one of ' f'{supported_types}',
```
Comment (jmarshall) [inline cpg_workflows/metamist.py:567]:
```suggestion
            f'{sequencing_group_id}: ERROR: "reads_type" is expected to be one of {supported_types}',
```

File: cpg_workflows/metamist.py [hunk 14]
```diff
@@ -645,24 +620,22 @@ def parse_reads(  # pylint: disable=too-many-return-statements
...
                     f'formatted. Expecting 2 entries per lane (R1 and R2 fastqs), '
                     f'but got {len(lane_pair)}. '
-                    f'Read data: {pprint.pformat(lane_pair)}'
+                    f'Read data: {pprint.pformat(lane_pair)}',
                 )
             if check_existence and not exists(lane_pair[0]['location']):
                 raise MetamistError(
-                    f'{sequencing_group_id}: ERROR: read 1 file does not exist: '
-                    f'{lane_pair[0]["location"]}'
+                    f'{sequencing_group_id}: ERROR: read 1 file does not exist: ' f'{lane_pair[0]["location"]}',
```
Comment (jmarshall) [inline cpg_workflows/metamist.py:627]:
```suggestion
                    f'{sequencing_group_id}: ERROR: read 1 file does not exist: {lane_pair[0]["location"]}',
```

File: cpg_workflows/metamist.py [hunk 15]
```diff
@@ -645,24 +620,22 @@ def parse_reads(  # pylint: disable=too-many-return-statements
...
                 )
             if check_existence and not exists(lane_pair[1]['location']):
                 raise MetamistError(
-                    f'{sequencing_group_id}: ERROR: read 2 file does not exist: '
-                    f'{lane_pair[1]["location"]}'
+                    f'{sequencing_group_id}: ERROR: read 2 file does not exist: ' f'{lane_pair[1]["location"]}',
```
Comment (jmarshall) [inline cpg_workflows/metamist.py:631]:
```suggestion
                    f'{sequencing_group_id}: ERROR: read 2 file does not exist: {lane_pair[1]["location"]}',
```

File: cpg_workflows/large_cohort/combiner.py (hunk 9 above)
Comment (illusional) [reply cpg_workflows/large_cohort/combiner.py]: I find / replaced these all locally in 17967aa

File: cpg_workflows/python_scripts/check_pedigree.py [hunk 16]
```diff
@@ -295,10 +283,7 @@ def print_contents(
     """
     if len(samples_df) < 400:
         samples_str = samples_df.to_string()
-        logging.info(
-            f'Somalier results, samples (based on {somalier_samples_fpath}):\n'
-            f'{samples_str}\n'
-        )
+        logging.info(f'Somalier results, samples (based on {somalier_samples_fpath}):\n' f'{samples_str}\n')
```
Comment (jmarshall) [inline cpg_workflows/python_scripts/check_pedigree.py:286]:
```suggestion
        logging.info(f'Somalier results, samples (based on {somalier_samples_fpath}):\n{samples_str}\n')
```

File: cpg_workflows/python_scripts/check_pedigree.py [hunk 17]
```diff
@@ -311,10 +296,7 @@ def print_contents(
                 'expected_relatedness',
             ]
         ].to_string()
-        logging.info(
-            f'Somalier results, sample pairs (based on {somalier_pairs_fpath}):\n'
-            f'{pairs_str}\n'
-        )
+        logging.info(f'Somalier results, sample pairs (based on {somalier_pairs_fpath}):\n' f'{pairs_str}\n')
```
Comment (jmarshall) [inline cpg_workflows/python_scripts/check_pedigree.py:299]:
```suggestion
        logging.info(f'Somalier results, sample pairs (based on {somalier_pairs_fpath}):\n{pairs_str}\n')
```

File: cpg_workflows/query_modules/seqr_loader_sv.py [hunk 18]
```diff
@@ -441,19 +420,9 @@ def _filter_sample_cn(i, g):
         #     lambda g: g.new_call | hl.is_defined(g.prev_num_alt)
         # ),
         samples_no_call=_genotype_filter_samples(lambda g: g.num_alt == -1),
-        samples_num_alt=hl.struct(
-            **{
-                '%i' % i: _genotype_filter_samples(_filter_num_alt(i))
-                for i in range(1, 3, 1)
-            }
-        ),
+        samples_num_alt=hl.struct(**{'%i' % i: _genotype_filter_samples(_filter_num_alt(i)) for i in range(1, 3, 1)}),
```
Comment (jmarshall) [inline cpg_workflows/query_modules/seqr_loader_sv.py:423]: For these ones, I am extremely embarrassed at what my line length suggestion has done…

File: cpg_workflows/resources.py [hunk 19]
```diff
@@ -169,10 +166,7 @@ def adjust_ncpu(self, ncpu: int) -> int:
         the nearest power of 2, not less than the minimal number of cores allowed.
         """
         if ncpu > self.max_ncpu:
-            raise ValueError(
-                f'Requesting more cores than available on {self.name} machine: '
-                f'{ncpu}>{self.max_ncpu}'
-            )
+            raise ValueError(f'Requesting more cores than available on {self.name} machine: ' f'{ncpu}>{self.max_ncpu}')
```
Comment (jmarshall) [inline cpg_workflows/resources.py:169]:
```suggestion
            raise ValueError(f'Requesting more cores than available on {self.name} machine: {ncpu}>{self.max_ncpu}')
```

File: cpg_workflows/jobs/trim.py (hunk 8 above)
Comment (jmarshall) [reply cpg_workflows/jobs/trim.py:93]: It's a bit awkward but I guess IMHO it's better than splitting the options from their arguments…

File: cpg_workflows/stages/gcnv.py [hunk 20]
```diff
@@ -421,17 +399,11 @@ def queue_jobs(self, cohort: Cohort, inputs: StageInput) -> StageOutput | None:
...
 
         # run strvctvre
-        strv_job.command(
-            f'python StrVCTVRE.py '
-            f'-i {input_vcf} '
-            f'-o temp.vcf '
-            f'-f vcf '
-            f'-p {phylop_in_batch}'
-        )
+        strv_job.command(f'python StrVCTVRE.py ' f'-i {input_vcf} ' f'-o temp.vcf ' f'-f vcf ' f'-p {phylop_in_batch}')
```
Comment (jmarshall) [inline cpg_workflows/stages/gcnv.py:406]: LOL

```suggestion
        strv_job.command(f'python StrVCTVRE.py -i {input_vcf} -o temp.vcf -f vcf -p {phylop_in_batch}')
```

File: cpg_workflows/stages/joint_genotyping.py [hunk 21]
```diff
@@ -58,19 +56,14 @@ def queue_jobs(self, cohort: Cohort, inputs: StageInput) -> StageOutput | None:
                 logging.error(f'Joint genotyping: could not find GVCF for {sgid}')
                 not_found_gvcfs.append(sgid)
         if not_found_gvcfs:
-            raise WorkflowError(
-                f'Joint genotyping: could not find {len(not_found_gvcfs)} '
-                f'GVCFs, exiting'
-            )
+            raise WorkflowError(f'Joint genotyping: could not find {len(not_found_gvcfs)} ' f'GVCFs, exiting')
```
Comment (jmarshall) [inline cpg_workflows/stages/joint_genotyping.py:59]:
```suggestion
            raise WorkflowError(f'Joint genotyping: could not find {len(not_found_gvcfs)} GVCFs, exiting')
```

File: cpg_workflows/status.py [hunk 22]
```diff
@@ -83,10 +86,7 @@ def complete_analysis_job(
         traceback.print_exc()
         raise
     else:
-        print(
-            f'Created Analysis(id={a_id}, type={analysis_type}, '
-            f'output={output}) in {project_name}'
-        )
+        print(f'Created Analysis(id={a_id}, type={analysis_type}, ' f'output={output}) in {project_name}')
```
Comment (jmarshall) [inline cpg_workflows/status.py:89]:
```suggestion
        print(f'Created Analysis(id={a_id}, type={analysis_type}, output={output}) in {project_name}')
```

File: cpg_workflows/workflow.py [hunk 23]
```diff
@@ -285,12 +280,11 @@ def _get(
...
                 f'Not found output from stage {stage.__name__}, required for stage '
                 f'{self.stage.name}. Is {stage.__name__} in the `required_stages`'
-                f'decorator? Available: {self._outputs_by_target_by_stage}'
+                f'decorator? Available: {self._outputs_by_target_by_stage}',
             )
         if not self._outputs_by_target_by_stage[stage.__name__].get(target.target_id):
             raise StageInputNotFoundError(
-                f'Not found output for {target} from stage {stage.__name__}, required '
-                f'for stage {self.stage.name}'
+                f'Not found output for {target} from stage {stage.__name__}, required ' f'for stage {self.stage.name}',
```
Comment (jmarshall) [inline cpg_workflows/workflow.py:287]:
```suggestion
                f'Not found output for {target} from stage {stage.__name__}, required for stage {self.stage.name}',
```

File: cpg_workflows/workflow.py [hunk 24]
```diff
@@ -629,73 +604,58 @@ def _get_action(self, target: TargetT) -> Action:
...
-                    f'{first_missing_path}'
+                    f'{first_missing_path}',
                 )
 
         if reusable and not first_missing_path:
             if target.forced:
                 logging.info(
-                    f'{self.name}: {target} [QUEUE] (can reuse, but forcing the target '
-                    f'to rerun this stage)'
+                    f'{self.name}: {target} [QUEUE] (can reuse, but forcing the target ' f'to rerun this stage)',
```
Comment (jmarshall) [inline cpg_workflows/workflow.py:651]:
```suggestion
                    f'{self.name}: {target} [QUEUE] (can reuse, but forcing the target to rerun this stage)',
```

File: cpg_workflows/workflow.py [hunk 25]
```diff
@@ -629,73 +604,58 @@ def _get_action(self, target: TargetT) -> Action:
...
                 )
                 return Action.QUEUE
             elif self.forced:
-                logging.info(
-                    f'{self.name}: {target} [QUEUE] (can reuse, but forcing the stage '
-                    f'to rerun)'
-                )
+                logging.info(f'{self.name}: {target} [QUEUE] (can reuse, but forcing the stage ' f'to rerun)')
```
Comment (jmarshall) [inline cpg_workflows/workflow.py:655]:
```suggestion
                logging.info(f'{self.name}: {target} [QUEUE] (can reuse, but forcing the stage to rerun)')
```

File: cpg_workflows/workflow.py [hunk 26]
```diff
@@ -629,73 +604,58 @@ def _get_action(self, target: TargetT) -> Action:
...
                 return Action.QUEUE
             else:
-                logging.info(
-                    f'{self.name}: {target} [REUSE] (expected outputs exist: '
-                    f'{expected_out})'
-                )
+                logging.info(f'{self.name}: {target} [REUSE] (expected outputs exist: ' f'{expected_out})')
```
Comment (jmarshall) [inline cpg_workflows/workflow.py:658]:
```suggestion
                logging.info(f'{self.name}: {target} [REUSE] (expected outputs exist: {expected_out})')
```

File: cpg_workflows/workflow.py [hunk 27]
```diff
@@ -1118,13 +1061,10 @@ def set_stages(
...
                 "Workflow config parameter 'only_stages' is incompatible with "
-                + "'first_stages', 'last_stages' and/or 'skip_stages'"
+                + "'first_stages', 'last_stages' and/or 'skip_stages'",
             )
 
-        logging.info(
-            f'End stages for the workflow "{self.name}": '
-            f'{[cls.__name__ for cls in requested_stages]}'
-        )
+        logging.info(f'End stages for the workflow "{self.name}": ' f'{[cls.__name__ for cls in requested_stages]}')
```
Comment (jmarshall) [inline cpg_workflows/workflow.py:1067]:
```suggestion
        logging.info(f'End stages for the workflow "{self.name}": {[cls.__name__ for cls in requested_stages]}')
```

File: cpg_workflows/workflow.py [hunk 28]
```diff
@@ -1160,10 +1100,7 @@ def set_stages(
                     newly_implicitly_added_d[reqstg.name] = reqstg
 
             if newly_implicitly_added_d:
-                logging.info(
-                    f'Additional implicit stages: '
-                    f'{list(newly_implicitly_added_d.keys())}'
-                )
+                logging.info(f'Additional implicit stages: ' f'{list(newly_implicitly_added_d.keys())}')
```
Comment (jmarshall) [inline cpg_workflows/workflow.py:1103]:
```suggestion
                logging.info(f'Additional implicit stages: {list(newly_implicitly_added_d.keys())}')
```

File: cpg_workflows/workflow.py [hunk 29]
```diff
@@ -1202,16 +1137,11 @@ def set_stages(
         if not (final_set_of_stages := [s.name for s in stages if not s.skipped]):
             raise WorkflowError('No stages to run')
 
-        logging.info(
-            f'Final list of stages after applying stage configuration options:\n'
-            f'{final_set_of_stages}'
-        )
+        logging.info(f'Final list of stages after applying stage configuration options:\n' f'{final_set_of_stages}')
```
Comment (jmarshall) [inline cpg_workflows/workflow.py:1140]:
```suggestion
        logging.info(f'Final list of stages after applying stage configuration options:\n{final_set_of_stages}')
```

File: pyproject.toml [hunk 30]
```diff
@@ -18,4 +19,189 @@ ignore_missing_imports = true
...
+
+section-order = ["future", "standard-library", "third-party", "hail", "cpg", "first-party", "local-folder"]
+
+[tool.ruff.lint.isort.sections]
+hail = [
+    "hail",
+    "hailtop",
+]
+# Adjust these for each repository, e.g., removing those that should be
+# local rather than CPG. Also fill in extend_skip below if there are any
```
Comment (jmarshall) [inline pyproject.toml:197]: The `extend_skip` bit was for isort, so can be replaced by a reference to `tool.ruff.extend-exclude`.

But basically the whole comment is only really useful in the template directory, so you could just remove it.

File: test/__init__.py [hunk 31]
```diff
@@ -16,8 +17,7 @@
 
 @runtime_checkable
 class IDictRepresentable(Protocol):
-    def as_dict(self) -> dict[str, Any]:
-        ...
+    def as_dict(self) -> dict[str, Any]: ...
```
Comment (jmarshall) [inline test/__init__.py:20]: I guess this one basically **is** a type annotation stub… (and I couldn't think of a good docstring for it)

File: test/jobs/somalier/test_relate.py [hunk 32]
```diff
@@ -249,15 +246,15 @@ def test_moves_somalier_output_to_expected_batch_resource_files(
 
         cmd = get_command_str(relate_j)
         assert re.search(
-            fr'mv related.pairs.tsv \${{BATCH_TMPDIR}}/.+\/output_pairs',
+            r'mv related.pairs.tsv \${BATCH_TMPDIR}/.+\/output_pairs',
```
Comment (jmarshall) [inline test/jobs/somalier/test_relate.py:249]: Apologies to the author of this, who now has to deal with half the lines saying `${{BATCH_TMPDIR}}` and the other half saying `${BATCH_TMPDIR}`… 🙃

File: test/jobs/test_align/test_bam_cram.py [hunk 33]
```diff
@@ -215,20 +206,15 @@ def test_indexes_bam_if_index_does_not_exist(self, tmp_path: Path):
...
-        file = re.escape(f'SAMPLE1.bam')
+        file = re.escape('SAMPLE1.bam')
 
         # Test sorts and indexes alignment input,
         assert re.search(
-            (
-                fr'samtools sort \${{BATCH_TMPDIR}}/inputs/\w+/{file}'
-                r' .* > \$BATCH_TMPDIR/sorted\.bam'
-            ),
+            (fr'samtools sort \${{BATCH_TMPDIR}}/inputs/\w+/{file}' r' .* > \$BATCH_TMPDIR/sorted\.bam'),
```
Comment (jmarshall) [inline test/jobs/test_align/test_bam_cram.py:213]: Here's another category of these (probably rare):

```suggestion
            (fr'samtools sort \${{BATCH_TMPDIR}}/inputs/\w+/{file} .* > \$BATCH_TMPDIR/sorted\.bam'),
```

**Comment by jmarshall** [comment]
I've looked over all of that now (my eyes glazed over a bit in the middle!) and modulo my usual rants I think it's just about ready to go.

I suggested the wider line length so that applying black wouldn't make edits to shorten any 88-120ish length lines we had, so that we could consider further what width windows we're all using and what's good here. However I (stupidly!) didn't anticipate that it would conversely make edits to join expressions over multiple lines that now fit in 120 characters! But IMHO that's made those lines more readable, so :+1:…

There's some `f'Hello ' f'world'` malarky to be fixed, a couple of other misc comments I've made (currently hidden by all the f-string comments, which I'll mark resolved as soon as they are), and there's I think three instances of two-list-items-on-a-line that I think we should address.

**Comment by jmarshall** [comment]
These two-list-items-on-a-line things are instances of

```python
['somecommand',
'-o', 'foo',
'-a', 'bar',
…
```

that it has reformatted to

```python
['somecommand',
'-o',
'foo',
'-a',
'bar',
…
```

I think that's ugly enough that we should stop it from doing that. Michael's suggestion is to use

```python
['somecommand',
*('-o', 'foo'),
*('-a', 'bar'),
…
```

which is probably the way to go.

**Comment by jmarshall** [comment]
I've pushed a commit for your consideration to tidy up the two-list-items-on-a-line thing.

One comment on _pyproject.toml_, and then this is :+1: from me.

File: pyproject.toml (hunk 30 above)
Comment (illusional) [reply pyproject.toml]: Yeah agreed, I've removed this comment

**Review by jmarshall — APPROVED** [review]

File: test/jobs/somalier/test_relate.py (hunk 32 above)
Comment (illusional) [reply test/jobs/somalier/test_relate.py:249]: placeholder in an a regex string, I don't love this haha

**Pull request merged**